src/core/electrostatics_magnetostatics/fft.cpp

```diff
@@ -482,8 +472,8 @@ void calc_2d_grid(int n, int grid[3]) {
 } // namespace
 
 int fft_init(const Utils::Vector3i &ca_mesh_dim, int const *ca_mesh_margin,
-             int *global_mesh_dim, double *global_mesh_off, int *ks_pnum,
-             fft_data_struct &fft, const Utils::Vector3i &grid,
+             int const *global_mesh_dim, double const *global_mesh_off,
+             int &ks_pnum, fft_data_struct &fft, const Utils::Vector3i &grid,
              const boost::mpi::communicator &comm) {
```

Comment:
I tried writing a C++ unit test for the FFT, but `MPI_Cart_coords()` fails with return value `MPI_ERR_TOPOLOGY`, and since this error code is not checked, the unit test continues and initializes an FFT plan with incorrect values that lead to integer overflow. I tried with
```c++
int main(int argc, char **argv) {
  auto mpi_env = mpi_init();
  Communication::init(mpi_env);
  return boost::unit_test::unit_test_main(init_unit_test, argc, argv);
}
```

and

```c++
int main(int argc, char **argv) {
  boost::mpi::environment mpi_env(argc, argv);
  return boost::unit_test::unit_test_main(init_unit_test, argc, argv);
}
```

without success.

This uncaught error code issue is unrelated to the issue with OpenMPI+UCX.

Reply:
I'd say this is a logic error, and MPI is trying to tell you that the topology you asked for can not be realized with the number of ranks. Is there any reason why you are using `Communication::init` and not just the lower level functions. I'd say calling into the core already partially defeats the purpose of the test.

Reply:
Btw. I *really* think the code should be moved out of the core before adding tests. We have to check that it is possible to run this standalone, otherwise we have to fix it. Logically, the fft code can node depend on core code. You don't have to know about espresso to do fft's ...

Reply:
More to your point: You should use the cartesian utility functions from utils/mpi, they have error check, and throw is there is a problem.

Reply:
Thanks, this sorts out the MPI error. The test draft is in 68b2bf081f7, I will not be working on it any further. I don't understand what the extra "padding" is for (i.e. the extension of buf_size, not the margin), I get a different FFT output depending on the number of MPI ranks, and the fftw3 on a 3D grid of cosines returns 0 instead of the the fundamental modes.